crates/ruff_server/src/message.rs

```diff
@@ -0,0 +1,63 @@
+use std::sync::OnceLock;
+
+use lsp_types::notification::Notification;
+
+use crate::server::ClientSender;
+
+static MESSENGER: OnceLock<ClientSender> = OnceLock::new();
+
+pub(crate) fn init_messenger(client_sender: &ClientSender) {
+    MESSENGER
+        .set(client_sender.clone())
+        .expect("messenger should only be initialized once");
+
+    // unregister any previously registered panic hook
+    let _ = std::panic::take_hook();
+
+    // When we panic, try to notify the client.
+    std::panic::set_hook(Box::new(move |panic_info| {
+        if let Some(messenger) = MESSENGER.get() {
+            let _ = messenger.send(lsp_server::Message::Notification(
+                lsp_server::Notification {
+                    method: lsp_types::notification::ShowMessage::METHOD.into(),
+                    params: serde_json::to_value(lsp_types::ShowMessageParams {
+                        typ: lsp_types::MessageType::ERROR,
+                        message: format!(
+                            "The Ruff language server exited with a panic: {panic_info}"
+                        ),
+                    })
+                    .unwrap_or_default(),
+                },
+            ));
+        }
```

Comment:
Could you extend your test plan to cover the panic hook? Just to make sure that it works.

Reply:
Done!